README.md
# data-retention-policy-linter

Check exported data-class retention policies for owners, durations, legal-hold
states, deletion-job references and conflicting environments.

**This tool deletes nothing.** It reads four JSON documents, writes a report, and
recommends. It opens no socket, runs no job, and has no clock — so it never says
that any particular record is now old enough to remove, only whether the written
policy for a class is consistent and unblocked.

**An active legal hold outranks every retention rule.** A class under one is never
recommended for deletion, whatever its duration says, and a deletion job that
covers it is reported as a conflict to suspend.

- **Repository:** [edilec/data-retention-policy-linter](https://github.com/edilec/data-retention-policy-linter)
- **Area:** Security & Privacy
- **License:** MIT
- Node ESM, `node >= 22`, no runtime and no development dependencies.

## Install and run

```sh
npx data-retention-policy-linter --root ./retention
```

```sh
data-retention-policy-linter --root examples/clean
data-retention-policy-linter --root examples/held --json | jq '.classes[] | {id, recommendation}'
```

stdout carries the JSON report and nothing else, so it can be piped straight into
a parser. The human summary and every diagnostic go to stderr, which means a
non-empty stderr on a successful run is correct rather than a symptom.

| Exit | Meaning |
| ---: | --- |
| `0` | The documents were linted and nothing contradicted them. |
| `1` | They were linted and at least one error-severity rule fired. |
| `2` | Invalid configuration (nothing on stdout), or evidence that could not be obtained (an `incomplete` report on stdout, never a `pass`). |

## Input

Four documents in one directory. Every one declares `"schemaVersion": "1"`, and an
unknown key anywhere is refused rather than ignored so a typo cannot disable a
check.

```
retention/
  classes.json    the data classes: id, owner, optional regulatory minimum
  policies.json   one retention period per class per environment
  holds.json      legal holds, each active or released
  jobs.json       deletion jobs and the classes they cover
```

```json
// classes.json
{ "schemaVersion": "1", "classes": [
  { "id": "billing.invoices", "owner": "finance-platform",
    "regulation": "Companies Act 2013, s.128",
    "regulatoryMinimum": { "value": 8, "unit": "year" } }
]}

// policies.json
{ "schemaVersion": "1", "policies": [
  { "class": "billing.invoices", "environment": "production",
    "retention": { "value": 8, "unit": "year" } }
]}

// holds.json
{ "schemaVersion": "1", "holds": [
  { "id": "matter-2031-discovery", "status": "active", "classes": ["support.transcripts"] }
]}

// jobs.json
{ "schemaVersion": "1", "jobs": [
  { "id": "nightly-expiry-sweep", "classes": ["billing.invoices"] }
]}
```

Each document can be given a name of its own with `--classes`, `--policies`,
`--holds` and `--jobs`; each is resolved inside `--root`, with the real path of
both sides compared, so a symbolic link planted in the tree is refused while a
legitimate file under a symlinked root is not.

### Durations are compared as explicit units

A duration is `{"value": <integer>, "unit": "day" | "week" | "month" | "year"}`. A
duration written as a string — `"30d"`, `"P7Y"` — is reported as unsupported and
the entry carrying it is not evaluated, because `"1m"` is a minute to one exporter
and a month to another.

Months and years are not fixed numbers of days, so each duration compiles to the
interval of days it can span (a month is 28–31; a year is 365–366) and a
comparison answers *less*, *greater*, *equal* or **ambiguous**. `1 month` against
`30 day` is ambiguous: in February the month is shorter and in March it is longer.
An ambiguous comparison is a finding, it makes the run `incomplete`, and the class
it concerns is left undecided rather than nudged into whichever answer is
convenient.

## Output

```json
{
  "schemaVersion": "1",
  "tool": "data-retention-policy-linter",
  "status": "fail",
  "summary": {
    "checked": 4, "errors": 1, "warnings": 0,
    "classes": 4, "policies": 8, "holds": 2, "activeHolds": 1, "jobs": 1,
    "environments": 2,
    "deletionRecommended": 3, "blockedByHold": 1, "blocked": 0, "undecided": 0
  },
  "environments": ["production", "warm-standby"],
  "classes": [
    {
      "id": "support.transcripts",
      "owner": "support-operations",
      "regulation": "internal-quality-policy-4",
      "regulatoryMinimum": { "value": 90, "unit": "day" },
      "environments": ["production", "warm-standby"],
      "retention": [{ "environment": "production", "value": 180, "unit": "day" }],
      "holds": ["matter-2031-discovery"],
      "activeHolds": ["matter-2031-discovery"],
      "deletionJobs": ["nightly-expiry-sweep"],
      "recommendation": "blocked-by-hold",
      "reasons": ["held-class-has-deletion-job", "legal-hold-active"]
    }
  ],
  "findings": [ /* ... */ ]
}
```

`recommendation` is one of `eligible`, `blocked`, `blocked-by-hold` or
`undecided`, in reverse order of precedence: a hold outranks everything, an
unknown outranks a known problem, and only `eligible` permits deletion.

The conflicts surfaced are: a retention period shorter than a declared regulatory
minimum; two environments declaring provably different durations for one class; a
deletion job referencing a class no policy declares; a class no deletion job
covers; and a class with no owner. The full catalog of fifty rules, with every
severity and every limit, is in
[`docs/retention-rules.md`](./docs/retention-rules.md).

## Examples

```sh
npm run example                                         # examples/clean,      exit 0
node bin/data-retention-policy-linter.mjs --root examples/broken      # exit 1
node bin/data-retention-policy-linter.mjs --root examples/held        # exit 1
node bin/data-retention-policy-linter.mjs --root examples/incomplete  # exit 2
```

- **`clean`** — four owned, policed, covered classes and one released hold.
  Nothing to report.
- **`broken`** — a policy below its regulatory minimum, two environments that
  disagree, a class with no owner, a class no job covers, and a job naming a class
  that no longer exists. Completes and fails.
- **`held`** — the acceptance case. `support.transcripts` is kept for 180 days
  against a 90-day minimum, is owned, is policed in both environments, and is
  covered by the nightly deletion job. Every retention rule in the file says it
  may be swept. It is under an active legal hold, so it is **not** recommended for
  deletion and the job that covers it is reported as a conflict to suspend.
- **`incomplete`** — a hold whose status this build does not implement, and two
  environments whose durations overlap without coinciding. Exits 2; nothing is
  recommended.

## Limits and non-goals

This tool reads four exported documents. Everything it says is a statement about
those documents, and the list below is what it deliberately does not do and
cannot conclude.

**It does not delete, schedule, trigger or connect.** There is no write path: the
only file-system surface the source imports is `{ readFile, realpath, stat }`, and
`test/read-only.test.mjs` snapshots the input tree byte for byte around a real run
and asserts that nothing moved. No socket is opened either, proved by a
module-resolution guard that refuses every network builtin, denies fetch and
socket connect, and runs over inert loopback-looking input without opening a
test listener. A URL in a
description is data, not an instruction to fetch it; a job named in a document is
data, not an instruction to run it.

**It cannot say a record is old enough to delete.** There is no clock in the
output path. `eligible` means "the written policy for this class is consistent and
unblocked", never "this data has expired". Deciding what has actually aged out is
the job of the system that holds the data.

**It cannot say whether any system obeys these documents.** A deletion job
declared here may not exist, may not run, or may cover something else. A pass says
the four files agree with one another, not that reality agrees with them.

**It cannot tell you the law.** A regulatory minimum is read from the class
catalog as a declared fact. The tool checks the policy against the minimum
somebody wrote down.

**It cannot tell you whether a hold is still in force.** `status` is read from the
export, so a hold released yesterday and exported last week still reads as active.
That is the safe direction to be wrong in, and it is why an unrecognised status is
refused rather than read as released.

**It cannot see data nobody classified.** The catalog is the world as far as this
tool is concerned. A store full of records that no class declares produces no
finding at all, which is the most important limitation on this list.

**It does not guess.** An unrecognised hold status, a duration written as a
string, a unit it does not implement, a comparison with two answers, a class name
a hold mentions and the catalog does not: each is reported as unsupported or
unknown and makes the run `incomplete`. None of them is resolved in the permissive
direction, and while hold evidence is incomplete no class in the run is
recommended for deletion at all — including classes no readable hold mentions,
because the hold nobody could read is exactly the one that would have stopped a
deletion.

**A pass is bounded by its limits.** Every bound — bytes, classes, policies,
holds, jobs, class references per entry, environments, findings, runtime — is

CHANGELOG.md
# Changelog

This project follows [Keep a Changelog](https://keepachangelog.com/en/1.1.0/)
and [Semantic Versioning](https://semver.org/spec/v2.0.0.html). Rule ids are
part of the public surface: renaming one is a breaking change and is recorded
here.

## [Unreleased]

### Added

- First implementation of `data-retention-policy-linter`: reads a data-class
  catalog, a set of retention policies, a set of legal holds and a set of
  deletion jobs, and reports what contradicts what. It deletes nothing,
  schedules nothing and recommends only.
- An active legal hold as the rule everything else is arranged around. A class
  under one is never recommended for deletion, whatever its retention period
  says; a deletion job covering it is reported as a conflict to suspend; and a
  held class is not reported as uncovered, because that would be advice to build
  a job for data under hold.
- Hold evidence treated as load-bearing. A refused hold entry, a hold status
  this build does not implement, or an unreadable `holds.json` means the set of
  held classes is not known, and while that is true no class in the run is
  recommended for deletion -- including classes no readable hold mentions.
- Durations compared as explicit units and never parsed out of a string. Each
  duration compiles to the interval of days it can span (a month is 28-31, a
  year 365-366), so a comparison answers less, greater, equal or *ambiguous*,
  and an ambiguous answer is a finding that makes the run `incomplete` rather
  than being rounded into whichever verdict is convenient.
- The conflicts the tool exists to surface: a retention period below a declared
  regulatory minimum, two environments declaring provably different durations
  for one class, a deletion job referencing a class no policy declares, a class
  no deletion job covers, and a class with no owner.
- A 50-rule catalog with one frozen `ruleId -> severity` table, documented in
  `docs/retention-rules.md` and pinned behaviourally by process exit code and
  literal error counts rather than by comparing three declarations.
- Enforced limits on bytes, classes, policies, holds, jobs, class references per
  entry, environments, duration values, findings and runtime, each reported by
  name when reached and each making the run `incomplete` rather than truncating
  silently. The time budget is re-checked *after* the evaluation loop returns,
  and every recommendation that loop reached is downgraded to `undecided` when
  it has been passed.
- A CLI with `--help`, `--json`, explicit input paths and the three documented
  exit codes; the JSON report on stdout alone.
- Examples for a clean, a failing, a legally held and an incomplete policy set.

### Security

- Read-only, and proved so: the shipped source imports `{ readFile, realpath,
  stat }` and nothing else from the file system, and a test snapshots the input
  tree byte for byte around a real run.
- No network access of any kind. Proved by a module-resolution guard that
  refuses every network builtin and by a live loopback listener whose address is
  planted in the input and never contacted.
- Path confinement resolves the real path of both the root and each input, so a
  symlink planted inside the root is refused while a legitimate file under a
  symlinked root is not.
- Strict UTF-8 decoding on every input, with no inference drawn from decoded
  text.
- Control (C0), DEL, C1, line/paragraph separator and bidi characters are
  stripped from every untrusted string that reaches output, identifiers and
  owners included; a value the tool refuses is described rather than reproduced.
  Two paths broke that rule and were fixed: an unsupported duration `unit` was
  quoted back verbatim under a docblock promising the opposite, and a stray
  object key was named at the document, entry and duration levels -- the entry
  message printing the key immediately after the sentence "no record, payload or
  credential field can reach this tool by accident". A unit is now described like
  every other refused value and a key is counted rather than named, so what an
  unknown key is called cannot change one byte of the report. The canary sweep,
  which had planted its placeholders only in a description, an owner, a class id
  and a whole-string retention, now plants them in a unit and in a key too.
- A parse failure does not quote the file it failed on. V8 writes
  `Unexpected token 'A', "AKIAIOSFODNN7EXAMPLE" is not valid JSON`, and
  `input-not-json` interpolated that message, so a retention file short enough
  to be nothing but a credential was reproduced in full on stdout -- the one
  claim above that stripping and truncation could not keep, since `excerpt` cuts

src/evaluate.mjs
/**
 * The evaluation: four compiled documents in, one row per data class out.
 *
 * Every row carries a `recommendation`, and the whole design of this module is
 * arranged around one rule:
 *
 * > **An active legal hold outranks every retention rule.** A class named by an
 * > active hold is never recommended for deletion, whatever its retention
 * > period says, whatever its regulatory minimum says, and whatever a deletion
 * > job already covering it says. A deletion job covering a held class is
 * > reported as a conflict to be suspended, not as a reason the hold is moot.
 *
 * The second rule follows from the first and is just as load-bearing: **hold
 * evidence that could not be read blocks every recommendation**. A hold entry
 * with a status this build does not implement, a hold naming something that is
 * not a name, an unreadable `holds.json` -- each of them means the set of held
 * classes is not known, and a class this run believes to be unheld might be
 * held by the hold nobody could read. So while hold evidence is incomplete, no
 * class in the run is recommended for deletion. The same reasoning, less
 * dramatically, applies to deletion-job evidence and to policies that could not
 * be attributed to a class.
 *
 * Nothing here deletes anything, schedules anything or contacts anything. The
 * output is a set of rows and a set of findings.
 */

import { formatDuration, relateDurations, satisfiesMinimum } from './duration.mjs'
import { byCodeUnit, excerpt } from './text.mjs'

/**
 * What a row's `recommendation` may be.
 *
 * Only `eligible` says deletion may proceed, and it says so about the *policy*,
 * never about a particular record: this tool has no clock, reads no data store
 * and knows nothing about how old anything is. See `README.md` for what that
 * does and does not establish.
 */
export const RECOMMENDATIONS = Object.freeze(['blocked', 'blocked-by-hold', 'eligible', 'undecided'])

/**
 * The vocabulary of `reasons`, deliberately separate from the rule catalog.
 *
 * A reason says why a class did not reach `eligible`; a rule id says what was
 * wrong with a document. They are usually related and they are not the same
 * thing -- one rule can produce several reasons and one reason can come from
 * several rules -- so keeping them apart stops a reader from mistaking the row
 * for a second copy of the findings list.
 */
export const REASONS = Object.freeze([
  'deletion-job-evidence-incomplete',
  'held-class-has-deletion-job',
  'legal-hold-active',
  'legal-hold-evidence-incomplete',
  'no-deletion-job',
  'no-retention-policy',
  'owner-missing',
  'policy-evidence-incomplete',
  'policy-unreadable',
  'regulatory-minimum-ambiguous',
  'regulatory-minimum-unreadable',
  'retention-below-minimum',
  'retention-conflict',
  'retention-conflict-ambiguous',
  'time-budget-exceeded',
])

const EVIDENCE_IDS = 8

/** A short, bounded list of ids for an evidence field. */
function listIds(ids) {
  const shown = ids.slice(0, EVIDENCE_IDS).map((id) => excerpt(id, 40))
  return ids.length > EVIDENCE_IDS ? `${shown.join(', ')} and ${ids.length - EVIDENCE_IDS} more` : shown.join(', ')
}

function countRecommendations(rows, wanted) {
  return rows.filter((row) => row.recommendation === wanted).length
}

function summarise(rows) {
  return {
    deletionRecommended: countRecommendations(rows, 'eligible'),
    blockedByHold: countRecommendations(rows, 'blocked-by-hold'),
    blocked: countRecommendations(rows, 'blocked'),
    undecided: countRecommendations(rows, 'undecided'),
  }
}

/**
 * Turn every conclusion this evaluation reached into `undecided`, except the
 * one that is already more restrictive than `undecided`.
 *
 * This exists for the case the report contract calls the cardinal sin: a budget
 * exhausted inside a loop, a `break` that falls through to the success branch,
 * and a confident claim written out on evidence the run never finished
 * gathering. The caller re-checks the budget *after* the loop and calls this
 * when it has been passed. `blocked-by-hold` is left alone on purpose -- it is
 * the strictest verdict there is, it cannot become `eligible`, and replacing it
 * would hide an active hold from the reader for no gain.
 */
export function downgradeRecommendations(result, reason) {
  for (const row of result.rows) {
    if (row.recommendation === 'blocked-by-hold') continue
    row.recommendation = 'undecided'
    if (!row.reasons.includes(reason)) {
      row.reasons.push(reason)
      row.reasons.sort(byCodeUnit)
    }
  }
  result.counts = summarise(result.rows)
  return result
}

/**
 * Compare four compiled documents.
 *
 * Returns `null` when the run could not be evaluated at all -- today that means
 * more environments than the limit allows. Otherwise returns the rows, the
 * environment list and the recommendation counts.
 */
export function evaluate(sink, names, compiled, limits, budget) {
  const { classes, policies, holds, jobs } = compiled

  const environmentSet = new Set()
  for (const policy of policies.entries) environmentSet.add(policy.environment)
  const environments = [...environmentSet].sort(byCodeUnit)

  if (environments.length > limits.maxEnvironments) {
    sink.add({
      file: names.policies,
      pointer: '/policies',
      ruleId: 'too-many-environments',
      message: `The policies name ${environments.length} environment(s), above the maxEnvironments limit of ${limits.maxEnvironments}; nothing was evaluated rather than some environments being compared and the rest ignored.`,
      suggestion: 'Raise --max-environments, or lint one group of environments at a time.',
    })
    return null
  }

  /**
   * Three completeness flags, each of which blocks every deletion
   * recommendation in the run while it is false.
   *
   * They are separate because they fail for different reasons and a reader
   * needs to know which, but they are used identically: an unknown is never
   * resolved in the permissive direction.
   */
  const holdEvidenceComplete = holds.refused === 0
  const jobEvidenceComplete = jobs.refused === 0
  const policyEvidenceComplete = policies.unattributedRefusals === 0

  if (!holdEvidenceComplete) {
    sink.add({
      file: names.holds,
      pointer: '/holds',
      ruleId: 'hold-coverage-unknown',
      message: `${holds.refused} legal hold(s) could not be read completely, so the set of classes under hold is not known. No class in this run is recommended for deletion while that is true -- including classes no hold this run could read mentions, because the hold nobody could read is exactly the one that would have stopped a deletion.`,
      suggestion: 'Fix the refused holds and re-run; until then treat every class as potentially held.',
    })
  }

  if (!jobEvidenceComplete) {
    sink.add({
      file: names.jobs,
      pointer: '/jobs',
      ruleId: 'job-coverage-unknown',
      message: `${jobs.refused} deletion job(s) could not be read completely, so which classes a job covers is not known. No class is reported as uncovered on this evidence, and none is recommended for deletion.`,
      suggestion: 'Fix the refused jobs and re-run.',
    })
  }

  if (!policyEvidenceComplete) {
    sink.add({
      file: names.policies,
      pointer: '/policies',
      ruleId: 'policy-coverage-unknown',
      message: `${policies.unattributedRefusals} retention policy entr(y/ies) were refused without naming a readable data class, so some class has a retention rule this run never read and there is no way to say which. No class is recommended for deletion on this evidence.`,
      suggestion: 'Fix the refused policies and re-run.',
    })
  }

  const holdsByClass = new Map()
  for (const hold of holds.entries) {
    if (hold.classes.length === 0) {
      sink.add({
        file: names.holds,
        pointer: `${hold.pointer}/classes`,
        ruleId: 'hold-covers-nothing',
        message: `Legal hold "${excerpt(hold.id, 80)}" names no data class, so it holds nothing here. It is recorded and it blocks nothing.`,
      })
      continue
    }
    for (const classId of hold.classes) {
      if (!classes.byId.has(classId)) {
        sink.add({
          file: names.holds,
          pointer: hold.classPointers.get(classId),
          ruleId: 'hold-class-unknown',
          message: `Legal hold "${excerpt(hold.id, 60)}" names data class "${excerpt(classId, 80)}", which the class catalog does not declare. The hold is recorded against nothing; this tool does not guess that a name it cannot find was meant to be one it can.`,
          suggestion: 'Declare the class, or correct the hold.',
        })
        continue
      }
      if (!holdsByClass.has(classId)) holdsByClass.set(classId, [])
      holdsByClass.get(classId).push(hold)
    }
  }

  const jobsByClass = new Map()
  for (const job of jobs.entries) {
    if (job.classes.length === 0) {
      sink.add({
        file: names.jobs,
        pointer: `${job.pointer}/classes`,
        ruleId: 'job-covers-nothing',
        message: `Deletion job "${excerpt(job.id, 80)}" names no data class, so it covers nothing here.`,
      })
      continue
    }
    for (const classId of job.classes) {
      if (!classes.byId.has(classId)) {
        sink.add({
          file: names.jobs,
          pointer: job.classPointers.get(classId),
          ruleId: 'job-class-unknown',
          message: `Deletion job "${excerpt(job.id, 60)}" names data class "${excerpt(classId, 80)}", which the class catalog does not declare. A job that deletes something nobody has classified is deleting on nobody's authority.`,
          suggestion: 'Declare the class, or correct the job.',
        })
        continue
      }
      if (!policies.byClass.has(classId) && !policies.refusedByClass.has(classId)) {
        sink.add({
          file: names.jobs,
          pointer: job.classPointers.get(classId),
          ruleId: 'job-class-unpolicied',
          message: `Deletion job "${excerpt(job.id, 60)}" covers data class "${excerpt(classId, 80)}", which no retention policy declares. The job would delete on a schedule nobody wrote down.`,
          suggestion: 'Declare a retention policy for this class, or remove it from the job.',
        })
      }
      if (!jobsByClass.has(classId)) jobsByClass.set(classId, [])
      jobsByClass.get(classId).push(job)
    }
  }

  for (const policy of policies.entries) {
    if (!classes.byId.has(policy.classId)) {
      sink.add({
        file: names.policies,
        pointer: `${policy.pointer}/class`,
        ruleId: 'policy-class-unknown',
        message: `This retention policy is written for data class "${excerpt(policy.classId, 80)}", which the class catalog does not declare; it was recorded against nothing and no class row mentions it.`,
        suggestion: 'Declare the class, or correct the policy.',
      })
    }
  }

  const rows = []

  for (const klass of classes.entries) {
    budget.check()

    const reasons = new Set()
    let blocked = false
    let unknown = false

    if (klass.ownerRefused) {
      reasons.add('owner-missing')
      blocked = true
    }
    if (klass.minimumRefused) {
      reasons.add('regulatory-minimum-unreadable')
      unknown = true
    }
    if (!holdEvidenceComplete) {
      reasons.add('legal-hold-evidence-incomplete')
      unknown = true
    }
    if (!jobEvidenceComplete) {
      reasons.add('deletion-job-evidence-incomplete')
      unknown = true
    }
    if (!policyEvidenceComplete) {
      reasons.add('policy-evidence-incomplete')
      unknown = true
    }

    const refusedPolicies = policies.refusedByClass.get(klass.id) ?? 0
    if (refusedPolicies > 0) {
      reasons.add('policy-unreadable')
      unknown = true
    }

    const classPolicies = (policies.byClass.get(klass.id) ?? [])
      .slice()
      .sort((left, right) => byCodeUnit(left.environment, right.environment))

    if (classPolicies.length === 0 && refusedPolicies === 0) {
      sink.add({
        file: names.classes,
        pointer: klass.pointer,
        ruleId: 'class-unpolicied',
        message: `Data class "${excerpt(klass.id, 100)}" has no retention policy in any environment, so there is no period after which anything may be deleted and no period before which anything must be kept.`,
        suggestion: 'Declare a retention policy for this class in each environment that stores it.',
      })
      reasons.add('no-retention-policy')
      blocked = true
    }

    if (classPolicies.length > 0 && environments.length > 1) {
      const present = new Set(classPolicies.map((policy) => policy.environment))
      const missing = environments.filter((environment) => !present.has(environment))
      if (missing.length > 0) {
        sink.add({
          file: names.classes,
          pointer: klass.pointer,
          ruleId: 'environment-coverage-partial',
          message: `Data class "${excerpt(klass.id, 80)}" has a retention policy in ${classPolicies.length} of the ${environments.length} environment(s) these policies name. A copy in an environment with no policy is a copy nobody has decided about.`,
          evidence: `no policy in: ${listIds(missing)}`,
          suggestion: 'Declare a policy for each environment that stores this class, or confirm it is not stored there.',
        })
      }
    }

    // Every unordered pair, because "A agrees with B" and "A agrees with C" do
    // not add up to "B agrees with C" once a duration can be an interval: one
    // month is consistent with both 28 days and 31 days, which are not
    // consistent with each other. The first conflicting pair and the first
    // undecidable pair in environment order are named; the rest would be the
    // same story told again.
    let conflict = null
    let ambiguous = null
    for (let left = 0; left < classPolicies.length; left += 1) {
      for (let right = left + 1; right < classPolicies.length; right += 1) {
        budget.check()
        const relation = relateDurations(classPolicies[left].retention, classPolicies[right].retention)
        if (relation === 'differs' && conflict === null) conflict = [classPolicies[left], classPolicies[right]]
        if (relation === 'ambiguous' && ambiguous === null) ambiguous = [classPolicies[left], classPolicies[right]]
      }
    }

    if (conflict !== null) {
      const [left, right] = conflict
      sink.add({
        file: names.classes,
        pointer: klass.pointer,
        ruleId: 'environment-duration-conflict',
        message: `Data class "${excerpt(klass.id, 80)}" is kept for provably different lengths of time in different environments: "${excerpt(left.environment, 40)}" keeps it ${formatDuration(left.retention)} and "${excerpt(right.environment, 40)}" keeps it ${formatDuration(right.retention)}.`,
        evidence: `${excerpt(left.environment, 40)}=${formatDuration(left.retention)}; ${excerpt(right.environment, 40)}=${formatDuration(right.retention)}`,
        suggestion: 'Settle on one retention period for this class, or record why the environments differ as two separate classes.',
      })
      reasons.add('retention-conflict')
      blocked = true
    }

    if (ambiguous !== null) {
      const [left, right] = ambiguous
      sink.add({
        file: names.classes,
        pointer: klass.pointer,
        ruleId: 'environment-duration-ambiguous',
        message: `Data class "${excerpt(klass.id, 80)}" is kept for ${formatDuration(left.retention)} in "${excerpt(left.environment, 40)}" and ${formatDuration(right.retention)} in "${excerpt(right.environment, 40)}"; those two periods overlap without coinciding, so whether the environments agree depends on the calendar. This build does not pick an answer.`,
        evidence: `${excerpt(left.environment, 40)}=${formatDuration(left.retention)}; ${excerpt(right.environment, 40)}=${formatDuration(right.retention)}`,
        suggestion: 'Restate both periods in the same unit so the comparison has one answer.',
      })
      reasons.add('retention-conflict-ambiguous')
      unknown = true
    }

    if (klass.regulatoryMinimum !== null) {
      for (const policy of classPolicies) {
        budget.check()
        const verdict = satisfiesMinimum(policy.retention, klass.regulatoryMinimum)
        const citation = klass.regulation === null ? '' : ` (${excerpt(klass.regulation, 80)})`
        if (verdict === 'below') {
          sink.add({
            file: names.policies,
            pointer: `${policy.pointer}/retention`,
            ruleId: 'retention-below-minimum',
            message: `"${excerpt(klass.id, 60)}" is kept for ${formatDuration(policy.retention)} in "${excerpt(policy.environment, 40)}", which is shorter than its declared regulatory minimum of ${formatDuration(klass.regulatoryMinimum)}${citation}.`,
            evidence: `retention=${formatDuration(policy.retention)}; minimum=${formatDuration(klass.regulatoryMinimum)}`,
            suggestion: 'Raise the retention period to the minimum, or correct the minimum.',
          })
          reasons.add('retention-below-minimum')
          blocked = true
        } else if (verdict === 'ambiguous') {
          sink.add({
            file: names.policies,
            pointer: `${policy.pointer}/retention`,
            ruleId: 'minimum-comparison-ambiguous',
            message: `"${excerpt(klass.id, 60)}" is kept for ${formatDuration(policy.retention)} in "${excerpt(policy.environment, 40)}" against a regulatory minimum of ${formatDuration(klass.regulatoryMinimum)}${citation}; those periods overlap without coinciding, so on some calendars the policy meets the minimum and on others it does not. This build reports that rather than choosing.`,
            evidence: `retention=${formatDuration(policy.retention)}; minimum=${formatDuration(klass.regulatoryMinimum)}`,
            suggestion: 'Restate the retention period and the minimum in the same unit.',
          })
          reasons.add('regulatory-minimum-ambiguous')
          unknown = true
        }
      }
    }

    const classHolds = holdsByClass.get(klass.id) ?? []
    const holdIds = classHolds.map((hold) => hold.id).sort(byCodeUnit)
    const activeHoldIds = classHolds.filter((hold) => hold.status === 'active').map((hold) => hold.id).sort(byCodeUnit)
    const jobIds = (jobsByClass.get(klass.id) ?? []).map((job) => job.id).sort(byCodeUnit)

    if (activeHoldIds.length > 0) {
      reasons.add('legal-hold-active')
      sink.add({
        file: names.classes,
        pointer: klass.pointer,
        ruleId: 'hold-active',
        message: `Data class "${excerpt(klass.id, 80)}" is under ${activeHoldIds.length} active legal hold(s). No deletion is recommended for it while that is true, whatever its retention period says; this run reports it as blocked-by-hold and deletes nothing.`,
        evidence: `holds: ${listIds(activeHoldIds)}`,
      })

      if (jobIds.length > 0) {
        reasons.add('held-class-has-deletion-job')
        sink.add({
          file: names.classes,
          pointer: klass.pointer,
          ruleId: 'hold-conflicts-with-job',
          message: `Data class "${excerpt(klass.id, 80)}" is under an active legal hold and is also covered by ${jobIds.length} deletion job(s); the hold outranks the retention rule, so those jobs must be suspended for this class until it is released. This tool reports the conflict and deletes nothing itself.`,
          evidence: `holds: ${listIds(activeHoldIds)}; jobs: ${listIds(jobIds)}`,
          suggestion: 'Suspend these jobs for this class, or release the hold if it no longer applies.',
        })
      }
    } else if (jobIds.length === 0 && jobEvidenceComplete) {
      // Suppressed for a held class on purpose: telling somebody that data
      // under legal hold has no deletion job is telling them to build one.
      sink.add({
        file: names.classes,
        pointer: klass.pointer,
        ruleId: 'class-uncovered-by-job',
        message: `No deletion job covers data class "${excerpt(klass.id, 100)}", so its retention period expires and nothing acts on it.`,
        suggestion: 'Add this class to a deletion job, or record why it is retained indefinitely.',
      })
      reasons.add('no-deletion-job')
      blocked = true
    }

    // Precedence, and the order matters. An active hold is the loudest
    // operational fact there is, so it is named first even when the class also
    // has conflicts; after that, an unknown outranks a known problem, because
    // "we could not tell" and "we can tell, and it is wrong" are different
    // things to hand a reader. Only the last branch permits deletion.
    let recommendation
    if (activeHoldIds.length > 0) recommendation = 'blocked-by-hold'
    else if (unknown) recommendation = 'undecided'
    else if (blocked) recommendation = 'blocked'
    else recommendation = 'eligible'

    rows.push({
      id: excerpt(klass.id, 120),
      owner: klass.owner === null ? null : excerpt(klass.owner, 120),
      regulation: klass.regulation === null ? null : excerpt(klass.regulation, 120),
      regulatoryMinimum: klass.regulatoryMinimum === null
        ? null
        : { value: klass.regulatoryMinimum.value, unit: klass.regulatoryMinimum.unit },
      environments: classPolicies.map((policy) => policy.environment).sort(byCodeUnit),
      retention: classPolicies.map((policy) => ({
        environment: policy.environment,
        value: policy.retention.value,
        unit: policy.retention.unit,
      })),
      holds: holdIds,
      activeHolds: activeHoldIds,
      deletionJobs: jobIds,
      recommendation,
      reasons: [...reasons].sort(byCodeUnit),
    })
  }

  rows.sort((left, right) => byCodeUnit(left.id, right.id))

  return { rows, environments, counts: summarise(rows) }
}
